fpga_interchange/arch_pack_clusters.cc

```diff
@@ -0,0 +1,585 @@
+            log_info("  - chainable pair: source %s - sink %s\n", IdString(chain_ports.cell_source).c_str(ctx),
+                     IdString(chain_ports.cell_sink).c_str(ctx));
+
+        if (cluster.cluster_cells_map.size() != 0)
+            log_info("  - cell port maps:\n");
+        for (auto cluster_cell : cluster.cluster_cells_map) {
+            log_info("    - cell: %s - port: %s\n", IdString(cluster_cell.cell).c_str(ctx),
+                     IdString(cluster_cell.port).c_str(ctx));
+        }
+    }
+}
+
+static bool check_cluster_cells_compatibility(CellInfo *old_cell, CellInfo *new_cell, pool<IdString> &exclude_nets)
+{
+    NPNR_ASSERT(new_cell->type == old_cell->type);
+    for (auto &new_port_pair : new_cell->ports) {
+        PortInfo new_port_info = new_port_pair.second;
+        PortInfo old_port_info = old_cell->ports.at(new_port_pair.first);
+
+        if (exclude_nets.count(new_port_info.net->name))
+            continue;
+
+        if (new_port_info.type != PORT_IN)
+            continue;
+
+        if (new_port_info.net != old_port_info.net)
+            return false;
+    }
+
+    return true;
+}
+
+void Arch::prepare_cluster(const ClusterPOD *cluster, uint32_t index)
+{
+    Context *ctx = getCtx();
+    IdString cluster_name(cluster->name);
+
+    pool<IdString> cluster_cell_types;
+    for (auto cell_type : cluster->root_cell_types)
+        cluster_cell_types.insert(IdString(cell_type));
+
+    // Find cluster roots
+    std::vector<CellInfo *> roots;
+    for (auto &cell : cells) {
+        CellInfo *ci = cell.second.get();
+
+        if (ci->cluster != ClusterId())
+            continue;
+
+        if (!cluster_cell_types.count(ci->type))
+            continue;
+
+        if (cluster->chainable_ports.size() == 0) {
+            ci->cluster.set(ctx, ci->name.str(ctx));
+            roots.push_back(ci);
+            continue;
+        }
+
+        // Only one type of dedicated interconnect is allowed.
+        auto chain_ports = cluster->chainable_ports[0];
+        IdString source_port(chain_ports.cell_source);
+        IdString sink_port(chain_ports.cell_sink);
+
+        PortRef driver = ci->ports[sink_port].net->driver;
+
+        if (driver.cell == nullptr || driver.port != source_port) {
+            // We hit a root cell
+            ci->cluster.set(ctx, ci->name.c_str(ctx));
+            roots.push_back(ci);
+
+            // Chained cells use dedicated connections, usually not exposed to the
+            // general interconnect resources. The port disconnection is required for
+            // sink ports which are connected to GND or VCC by default, which are not
+            // reachable due to the fixed dedicated interconnect.
+            // E.g.: The CI input of carry chains in 7series corresponds to the CIN bel port,
+            //       which can only be connected to the COUT output of the tile below.
+            disconnect_port(ctx, ci, sink_port);
```

Comment:
it's notable that "CIN connected to 0" and "CIN connected to 1" are cases that need to be told apart, both being supported but with distinct FASM features: https://github.com/SymbiFlow/prjxray-db/blob/c28433840752ff0da5b8353bfdeafed24148aa38/artix7/segbits_clbll_l.db#L674-L675

I guess the  plan is to determine this from the logical netlist during FASM generation, as the connectivity in the physical netlist will be lost?

Reply:
I think this is an interesting case where the CIN and CYINIT BEL inputs are actually physically the same, but their connectivity is expressed differently for P&R purposes. 

If the `CYINIT` cell port is emitted by yosys, than there would be no problems and the port will not get disconnected, because the `PRECYINIT` routing BEL is connected to the general routing. The problem is that the `GND` net is assigned to the `CI` cell port as well, and this cell port is mapped to the `CIN` bel port, which does not have connections to `GND` (at least from a P&R perspective).

The root of a chain, all the `C0`, `C1`, `AX` connections can be present for `CYINIT`, and the information preserved in the physical netlist as well.

Reply:
Yeah, that seems reasonable. We might need something somewhere to move constants from `CIN` to `CYINIT` if this proves to be a problem.